provision/lib/Warewulf/Provision/Dhcp/Dnsmasq.pm

```diff
@@ -0,0 +1,379 @@
+# Copyright (c) 2019 Stephan Weinberger <sweinberger@ubimet.com>
+#
+
+package Warewulf::Provision::Dhcp::Dnsmasq;
+
+use Warewulf::ACVars;
+use Warewulf::Logger;
+use Warewulf::Provision;
+use Warewulf::Provision::Dhcp;
+use Warewulf::DataStore;
+use Warewulf::Network;
+use Warewulf::SystemFactory;
+use Warewulf::Util;
+use Warewulf::Provision::Tftp;
+use Socket;
+
+our @ISA = ('Warewulf::Provision::Dhcp');
+
+=head1 NAME
+
+Warewulf::Provision::Dhcp::Dnsmasq - Warewulf's Dnsmasq interface.
+
+=head1 ABOUT
+
+
+=head1 SYNOPSIS
+
+    use Warewulf::Provision::Dhcp::Dnsmasq;
+
+    my $obj = Warewulf::Provision::Dhcp::Dnsmasq->new();
+
+
+=head1 METHODS
+
+=over 12
+
+=cut
+
+=item new()
+
+The new constructor will create the object that references configuration the
+stores.
+
+=cut
+
+sub
+new($$)
+{
+    my $proto = shift;
+    my $class = ref($proto) || $proto;
+    my $self = ();
+
+    $self = {};
+
+    bless($self, $class);
+
+    return $self->init(@_);
+}
+
+
+sub
+init()
+{
+    my $self = shift;
+    my $config = Warewulf::Config->new("provision.conf");
+
+    my @files = ('/etc/dnsmasq.conf');
+
+    if (my $file = $config->get("dhcpd config file")) {
+        &dprint("Using the DHCPD configuration file as defined by provision.conf\n");
+        if ($file =~ /^([a-zA-Z0-9_\-\.\/]+)$/) {
+            $self->set("FILE", $1);
+        } else {
+            &eprint("Illegal characters in path: $file\n");
+        }
+
+    }
+
+    if (! $self->get("FILE")) {
+        # Check if /etc/dnsmasq.d/ exists
+        if (-d "/etc/dnsmasq.d") {
+            $self->set("FILE", "/etc/dnsmasq.d/warewulf.conf");
+        }
+    }
+
+    if (! $self->get("FILE")) {
+        # Else look to see if we can find an existing dnsmasq.conf file
+        foreach my $file (@files) {
+            if ($file =~ /^([a-zA-Z0-9_\-\.\/]+)$/) {
+                my $file_clean = $1;
+                if (-f $file_clean) {
+                    $self->set("FILE", $file_clean);
+                    &dprint("Found DHCPD configuration file: $file_clean\n");
+                }
+            } else {
+                &eprint("Illegal characters in path: $file\n");
+            }
+        }
+        # If we couldn't find one, lets set it to a sane default and hope for the best
+        if (! $self->get("FILE")) {
+            &dprint("Probing dhcpd looking for a default config path\n");
+            if (-x "/usr/sbin/dnsmasq") {
+                open(CONF, "strings /usr/sbin/dnsmasq | grep '/dnsmasq.conf' | grep '^/etc/' |");
```

Comment:
Let's not go as far running strings on the binary to auto-discover the path for dnsmasq.conf.

Reply:
Instead lets just bail out at this point.

Reply:
Well, that was a copy from the Isc module. I feel that to be quite "brutal" there as well...

Reply:
Ah wonderful... Well let's not repeat what we did for ISC.